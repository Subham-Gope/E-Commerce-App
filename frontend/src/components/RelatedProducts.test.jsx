import { describe, test, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import ShopContext from "../context/ShopContext";
import RelatedProducts from "./RelatedProducts";

// mock Products
const mockProducts = [
  {
    _id: "1",
    name: "Black Tee",
    price: 25,
    image: ["black.jpg"],
    category: "clothing",
    subCategory: "tshirt",
  },
  {
    _id: "2",
    name: "White Tee",
    price: 30,
    image: ["white.jpg"],
    category: "clothing",
    subCategory: "tshirt",
  },
  {
    _id: "3",
    name: "Grey Tee",
    price: 20,
    image: ["grey.jpg"],
    category: "clothing",
    subCategory: "tshirt",
  },
  {
    _id: "4",
    name: "Green Tee",
    price: 28,
    image: ["green.jpg"],
    category: "clothing",
    subCategory: "tshirt",
  },
  {
    _id: "5",
    name: "Blue Tee",
    price: 22,
    image: ["blue.jpg"],
    category: "clothing",
    subCategory: "tshirt",
  },
  {
    _id: "6",
    name: "Denim Jacket",
    price: 60,
    image: ["jacket.jpg"],
    category: "clothing",
    subCategory: "jacket",
  },
];

// render with context function
const renderWithContext = (ui, value) => {
  return render(
    <MemoryRouter>
      <ShopContext.Provider value={value}>{ui}</ShopContext.Provider>
    </MemoryRouter>
  );
};

describe("RelatedProducts.jsx", () => {
  test("renders 5 related items, filters by category + subCategory", () => {
    // rendering and settin up the snapshot matching
    const { container } = renderWithContext(
      <RelatedProducts category="clothing" subCategory="tshirt" />,
      {
        products: mockProducts,
      }
    );
    expect(container).toMatchSnapshot();

    // checking the title
    expect(screen.getByText(/related/i)).toBeInTheDocument();

    // checking the displayed products using the link tag , should be 5 in total
    const links = screen.getAllByRole("link");
    expect(links).toHaveLength(5);

    //checking if 6th product is not displayed
    expect(screen.queryByText(/Denim Jacket/i)).toBeNull();
  });
});
