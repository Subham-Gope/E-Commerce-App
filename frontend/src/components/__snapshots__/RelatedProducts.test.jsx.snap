// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RelatedProducts.jsx > renders 5 related items, filters by category + subCategory 1`] = `
<div>
  <div
    class="my-24"
  >
    <div
      class="text-center text-3xl py-2"
    >
      <div
        class="inline-flex gap-2 items-center mb-3"
      >
        <p
          class="text-gray-500"
        >
          RELATED
           
          <span
            class="text-gray-700 font-medium"
          >
            PRODUCTS
          </span>
        </p>
        <p
          class="w-8 sm:w-12 h-[1px] sm:h-[2px] bg-gray-700"
        />
      </div>
    </div>
    <div
      class="grid grid-cols-2 sm:grid-cols-4 lg:grid-cols-5 gap-4 gap-y-6"
    >
      <a
        class="text-gray-700 cursor-pointer"
        data-discover="true"
        href="/product/1"
      >
        <div
          class="overflow-hidden"
        >
          <img
            alt="product image"
            class="hover:scale-110 transition ease-in-out"
            src="black.jpg"
          />
        </div>
        <p
          class="pt-3 pb-1 text-sm"
        >
          Black Tee
        </p>
        <p
          class="text-sm font-medium"
        >
           undefined 25
        </p>
      </a>
      <a
        class="text-gray-700 cursor-pointer"
        data-discover="true"
        href="/product/2"
      >
        <div
          class="overflow-hidden"
        >
          <img
            alt="product image"
            class="hover:scale-110 transition ease-in-out"
            src="white.jpg"
          />
        </div>
        <p
          class="pt-3 pb-1 text-sm"
        >
          White Tee
        </p>
        <p
          class="text-sm font-medium"
        >
           undefined 30
        </p>
      </a>
      <a
        class="text-gray-700 cursor-pointer"
        data-discover="true"
        href="/product/3"
      >
        <div
          class="overflow-hidden"
        >
          <img
            alt="product image"
            class="hover:scale-110 transition ease-in-out"
            src="grey.jpg"
          />
        </div>
        <p
          class="pt-3 pb-1 text-sm"
        >
          Grey Tee
        </p>
        <p
          class="text-sm font-medium"
        >
           undefined 20
        </p>
      </a>
      <a
        class="text-gray-700 cursor-pointer"
        data-discover="true"
        href="/product/4"
      >
        <div
          class="overflow-hidden"
        >
          <img
            alt="product image"
            class="hover:scale-110 transition ease-in-out"
            src="green.jpg"
          />
        </div>
        <p
          class="pt-3 pb-1 text-sm"
        >
          Green Tee
        </p>
        <p
          class="text-sm font-medium"
        >
           undefined 28
        </p>
      </a>
      <a
        class="text-gray-700 cursor-pointer"
        data-discover="true"
        href="/product/5"
      >
        <div
          class="overflow-hidden"
        >
          <img
            alt="product image"
            class="hover:scale-110 transition ease-in-out"
            src="blue.jpg"
          />
        </div>
        <p
          class="pt-3 pb-1 text-sm"
        >
          Blue Tee
        </p>
        <p
          class="text-sm font-medium"
        >
           undefined 22
        </p>
      </a>
    </div>
  </div>
</div>
`;
